pkg/provisioner/ironic/ironic.go

```diff
@@ -1523,5 +1536,53 @@ func (p *ironicProvisioner) IsReady() (result bool, err error) {
 	p.log.Info("verifying ironic provisioner dependencies")
 
 	checker := newIronicDependenciesChecker(p.client, p.inspector, p.log)
-	return checker.IsReady()
+	result, err = checker.IsReady()
+	if result {
+		result, err = p.checkMaxProvisioningHosts()
+	}
+
+	return
+}
+
+func (p *ironicProvisioner) checkMaxProvisioningHosts() (bool, error) {
+
+	current := 0
+
+	pager := nodes.List(p.client, nodes.ListOpts{
+		Fields: []string{"uuid,provision_state,driver_internal_info"},
+	})
+	if pager.Err != nil {
+		return false, pager.Err
+	}
+
+	page, err := pager.AllPages()
+	if err != nil {
+		return false, err
+	}
+
+	allNodes, err := nodes.ExtractNodes(page)
+	if err != nil {
+		return false, err
+	}
+
+	for _, node := range allNodes {
+
+		if _, ok := node.DriverInternalInfo["agent_url"]; !ok {
+			continue
+		}
+
+		switch node.ProvisionState {
```

Comment:
Would looking at the target provisioning state instead/as well provide a faster response to changes?

Reply:
Unfortunately it didn't :/. In my tests also TargetState usually points just to `manageable` or `active` states

Reply:
At the start of provisioning we are usually in manageable state (not always after #716, but always for the first deployment). So if we're seeing e.g. current state is manageable but target state is available, that machine is probably about to be provisioned. Ditto for current state is available and target state is active.
I assume the target state must be updated synchronously, so it really should be available whenever we look.

I suspect the real problem here is that there's a large delay between the point where you're making a go/no-go decision (on entry to the Provisioning state) and when we actually ask Ironic to go to active (not until after cleaning). And now it's even more fun because a node in the available state may _or may not_ represent a Host in the provisioning state.

I think the solution to this is the Preparing state I proposed [on the mailing list](https://groups.google.com/g/metal3-dev/c/N-EpqQn94bw). That way all nodes will be in available before starting provisioning, so we'd only have to look for those with target state active.

Reply:
@dtantsur can you please confirm such behavior? From my local tests the state doesn't seem to be update synchronously, neither the target_provision_state

Reply:
> current state is manageable but target state is available

To nitpick a bit: this is not possible. So called stable states always have target state None. Stable states are states where a node simply stays without external input (BMO is external input in this context). This includes ``manageable``, ``available``, ``active``. Non-empty target state appears in so called transient states: ``wait call-back``, all states ending with ``..ing`` or ``.. wait``.

Now let me check the code to provide a more targeted answer.